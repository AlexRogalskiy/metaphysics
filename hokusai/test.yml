---
version: '2'
services:
  metaphysics:
    command: yarn test
    extends:
      file: common.yml
      service: metaphysics
    environment:
    - NODE_ENV=test
    env_file: ../.env.test
