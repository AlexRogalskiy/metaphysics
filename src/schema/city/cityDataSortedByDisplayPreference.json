[
  {
    "slug": "new-york-ny-usa",
    "name": "New York",
    "coordinates": { "lat": 40.71, "lng": -74.01 }
  },
  {
    "slug": "los-angeles-ca-usa",
    "name": "Los Angeles",
    "coordinates": { "lat": 34.05, "lng": -118.24 }
  },
  {
    "slug": "london-united-kingdom",
    "name": "London",
    "coordinates": { "lat": 51.51, "lng": -0.13 }
  },
  {
    "slug": "berlin-germany",
    "name": "Berlin",
    "coordinates": { "lat": 52.52, "lng": 13.4 }
  },
  {
    "slug": "paris-france",
    "name": "Paris",
    "coordinates": { "lat": 48.86, "lng": 2.35 }
  },
  {
    "slug": "hong-kong-hong-kong",
    "name": "Hong Kong",
    "coordinates": { "lat": 22.3, "lng": 114.2 }
  }
]
