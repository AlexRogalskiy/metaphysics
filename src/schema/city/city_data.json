{
  "new-york-ny-usa": {
    "slug": "new-york-ny-usa",
    "name": "New York",
    "coordinates": { "lat": 40.71, "lng": -74.01 }
  },
  "london-united-kingdom": {
    "slug": "london-united-kingdom",
    "name": "London",
    "coordinates": { "lat": 51.51, "lng": -0.13 }
  },
  "los-angeles-ca-usa": {
    "slug": "los-angeles-ca-usa",
    "name": "Los Angeles",
    "coordinates": { "lat": 34.05, "lng": -118.24 }
  },
  "paris-france": {
    "slug": "paris-france",
    "name": "Paris",
    "coordinates": { "lat": 48.86, "lng": 2.35 }
  },
  "berlin-germany": {
    "slug": "berlin-germany",
    "name": "Berlin",
    "coordinates": { "lat": 52.52, "lng": 13.4 }
  },
  "hong-kong-hong-kong": {
    "slug": "hong-kong-hong-kong",
    "name": "Hong Kong",
    "coordinates": { "lat": 22.4, "lng": 114.11 }
  }
}
