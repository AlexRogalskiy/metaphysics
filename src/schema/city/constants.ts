export const LOCAL_DISCOVERY_RADIUS_KM = 75
