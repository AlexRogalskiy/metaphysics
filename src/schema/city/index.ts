import {
  GraphQLInt,
  GraphQLList,
  GraphQLNonNull,
  GraphQLObjectType,
  GraphQLString,
} from "graphql"

import { LatLngType } from "../location"
import { showConnection } from "schema/show"
import PartnerShowSorts from "schema/sorts/partner_show_sorts"
import Fair from "schema/fair"
import FairSorts from "schema/sorts/fair_sorts"
import EventStatus from "schema/input_fields/event_status"

import cityData from "./city_data.json"
import { pageable } from "relay-cursor-paging"
import { connectionFromArray } from "graphql-relay"

const LOCAL_DISCOVERY_RADIUS_KM = 75

const CityType = new GraphQLObjectType({
  name: "City",
  fields: {
    slug: {
      type: GraphQLString,
    },
    name: {
      type: GraphQLString,
    },
    coordinates: {
      type: LatLngType,
    },
    shows: {
      type: showConnection,
      args: pageable({
        size: {
          type: GraphQLInt,
        },
        sort: PartnerShowSorts,
        status: EventStatus,
      }),
      resolve: (city, args, _context, { rootValue: { showsLoader } }) => {
        const gravityOptions = {
          ...args,
          displayable: true,
          near: `${city.coordinates.lat},${city.coordinates.lng}`,
          max_distance: LOCAL_DISCOVERY_RADIUS_KM,
        }

        return showsLoader(gravityOptions).then(response => {
          return connectionFromArray(response, args)
        })
      },
    },
    fairs: {
      type: new GraphQLList(Fair.type),
      args: {
        size: { type: GraphQLInt },
        sort: FairSorts,
        status: EventStatus,
      },
      resolve: (obj, args, _context, { rootValue: { fairsLoader } }) => {
        const gravityOptions = {
          ...args,
          near: `${obj.coordinates.lat},${obj.coordinates.lng}`,
          max_distance: LOCAL_DISCOVERY_RADIUS_KM,
        }
        return fairsLoader(gravityOptions)
      },
    },
  },
})

export const City = {
  type: CityType,
  description: "A city-based entry point for local discovery",
  args: {
    slug: {
      type: new GraphQLNonNull(GraphQLString),
      description:
        "A slug for the city, conforming to Gravity's city slug naming conventions",
    },
  },
  resolve: (_obj, args) => {
    return lookupCity(args.slug)
  },
}

const lookupCity = slug => {
  if (!cityData.hasOwnProperty(slug)) {
    throw new Error(
      `City ${slug} not found in: ${Object.keys(cityData).join(", ")}`
    )
  }
  return cityData[slug]
}
