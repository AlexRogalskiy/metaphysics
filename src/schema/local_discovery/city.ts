import {
  GraphQLFloat,
  GraphQLInt,
  GraphQLList,
  GraphQLNonNull,
  GraphQLObjectType,
  GraphQLString,
} from "graphql"

import Show from "schema/show"
import EventStatus from "schema/input_fields/event_status"
import PartnerShowSorts from "schema/sorts/partner_show_sorts"

import cityData from "./city_data.json"

const LOCAL_DISCOVERY_RADIUS_KM = 75

const LatLngType = new GraphQLObjectType({
  name: "LatLng",
  fields: {
    lat: {
      type: GraphQLFloat,
      resolve: obj => {
        return obj.lat
      },
    },
    lng: {
      type: GraphQLFloat,
      resolve: obj => {
        return obj.lng
      },
    },
  },
})

const LocalDiscoveryCityType = new GraphQLObjectType({
  name: "LocalDiscoveryCity",
  fields: {
    slug: {
      type: GraphQLString,
      resolve: obj => {
        return obj.slug
      },
    },
    name: {
      type: GraphQLString,
      resolve: obj => {
        return obj.name
      },
    },
    coordinates: {
      type: LatLngType,
      resolve: obj => {
        return obj.coordinates
      },
    },
    shows: {
      type: new GraphQLList(Show.type),
      args: {
        size: { type: GraphQLInt },
        sort: PartnerShowSorts,
        status: EventStatus,
      },
      resolve: (obj, args, _context, { rootValue: { showsLoader } }) => {
        let gravityOptions = {
          ...args,
          displayable: true,
          near: `${obj.coordinates.lat},${obj.coordinates.lng}`,
          max_distance: LOCAL_DISCOVERY_RADIUS_KM,
        }

        // here is where we could fetch non-partner shows and
        // merge them in with partner shows.
        //
        // but how to deal with sorting across two disparate
        // collections?
        //
        // then again, if we model non-partner shows akin to
        // current reference shows, this problem may be solvable

        return showsLoader(gravityOptions)
      },
    },
  },
})

export const LocalDiscoveryCity = {
  type: LocalDiscoveryCityType,
  description: "A city-based entry point for local discovery",
  args: {
    slug: {
      type: new GraphQLNonNull(GraphQLString),
      description:
        "A slug for the city, conforming to Gravity's city slug naming conventions",
    },
  },
  resolve: (_obj, args, _context, _info) => {
    return lookupCity(args.slug)
  },
}

const lookupCity = slug => {
  if (!cityData.hasOwnProperty(slug)) {
    throw new Error(
      `City ${slug} not found in : ${Object.keys(cityData).join(", ")}`
    )
  }
  return cityData[slug]
}
