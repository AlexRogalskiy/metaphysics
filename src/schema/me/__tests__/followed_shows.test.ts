/* eslint-disable promise/always-return */
import { runAuthenticatedQuery } from "test/utils"
import gql from "lib/gql"
import cityData from "../../../schema/city/cityDataSortedByDisplayPreference.json"
import { LOCAL_DISCOVERY_RADIUS_KM } from "../../../schema/city/constants"

const BASE_GRAVITY_ARGS = { size: 10, offset: 0, total_count: true }
const stubResolver = () => Promise.resolve({ body: [], headers: {} })

const generate_query = (params = `(first: 10)`) =>
  gql`
    {
      me {
        followsAndSaves {
          shows${params} {
            edges {
              node {
                id
              }
            }
          }
        }
      }
    }
  `

describe("returns followed shows for a user", () => {
  let followedShowsLoader

  beforeEach(() => {
    followedShowsLoader = jest.fn(stubResolver)
  })

  it("generates a predictable URL with no parameters", async () => {
    const query = generate_query()
    await runAuthenticatedQuery(query, { followedShowsLoader })
    expect(followedShowsLoader).toHaveBeenCalledWith(BASE_GRAVITY_ARGS)
  })

  describe("respects optional city parameter", () => {
    const generate_city_args = city =>
      !city
        ? BASE_GRAVITY_ARGS
        : {
            ...BASE_GRAVITY_ARGS,
            near: `${city.coordinates.lat},${city.coordinates.lng}`,
            max_distance: LOCAL_DISCOVERY_RADIUS_KM,
          }

    const buildTestData = city => ({
      name: city.name,
      query: generate_query(`(first: 10, city: "${city.slug}")`),
      expectedArgs: generate_city_args(city),
    })

    cityData.map(buildTestData).forEach(({ name, query, expectedArgs }) => {
      // I totally grant that this is a bit weird
      // but the behavior under test is tightly coupled
      // to the contents of the cityData file. If a change
      // to that file breaks this test I want to know about it,
      // and this will tell us.

      // When Jest hits this point in the code it will treat this almost
      // like a macro - we'll have a new test generated for each city,
      // you'll see the test results enumerated as if we'd done this manually.
      it.call(null, `Handles City: ${name}`, async () => {
        await runAuthenticatedQuery(query, { followedShowsLoader })
        expect(followedShowsLoader).toHaveBeenCalledWith(expectedArgs)
      })
    })

    it("throws an error if presented with an invalid city slug", async () => {
      const { query } = buildTestData({
        slug: "invalid-city-location",
        coordinates: { lat: 0, lng: 0 },
      })

      try {
        await runAuthenticatedQuery(query, { followedShowsLoader })
        // shouldn't get here - if we did then the error wasn't thrown.
        expect(true).toBeFalsy()
      } catch (e) {
        expect(e).toBeDefined()
      }
    })
  })
})
