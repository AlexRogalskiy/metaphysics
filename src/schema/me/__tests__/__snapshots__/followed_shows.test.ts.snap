// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`returns followed shows for a user relies on the state of cityData 1`] = `
Array [
  Object {
    "coordinates": Object {
      "lat": 40.713,
      "lng": -74.006,
    },
    "maxBounds": Object {
      "ne": Object {
        "lat": 41.272,
        "lng": -73.266,
      },
      "sw": Object {
        "lat": 40.144,
        "lng": -74.754,
      },
    },
    "name": "New York",
    "slug": "new-york-ny-usa",
  },
  Object {
    "coordinates": Object {
      "lat": 34.052,
      "lng": -118.244,
    },
    "maxBounds": Object {
      "ne": Object {
        "lat": 34.644,
        "lng": -117.521,
      },
      "sw": Object {
        "lat": 33.453,
        "lng": -118.959,
      },
    },
    "name": "Los Angeles",
    "slug": "los-angeles-ca-usa",
  },
  Object {
    "coordinates": Object {
      "lat": 51.507,
      "lng": -0.128,
    },
    "maxBounds": Object {
      "ne": Object {
        "lat": 52.127,
        "lng": 0.862,
      },
      "sw": Object {
        "lat": 50.895,
        "lng": -1.118,
      },
    },
    "name": "London",
    "slug": "london-united-kingdom",
  },
  Object {
    "coordinates": Object {
      "lat": 52.52,
      "lng": 13.405,
    },
    "maxBounds": Object {
      "ne": Object {
        "lat": 53.159,
        "lng": 14.452,
      },
      "sw": Object {
        "lat": 51.882,
        "lng": 12.352,
      },
    },
    "name": "Berlin",
    "slug": "berlin-germany",
  },
  Object {
    "coordinates": Object {
      "lat": 48.857,
      "lng": 2.352,
    },
    "maxBounds": Object {
      "ne": Object {
        "lat": 49.377,
        "lng": 3.139,
      },
      "sw": Object {
        "lat": 48.339,
        "lng": 1.561,
      },
    },
    "name": "Paris",
    "slug": "paris-france",
  },
  Object {
    "coordinates": Object {
      "lat": 22.282,
      "lng": 114.158,
    },
    "maxBounds": Object {
      "ne": Object {
        "lat": 22.938,
        "lng": 114.891,
      },
      "sw": Object {
        "lat": 21.66,
        "lng": 113.509,
      },
    },
    "name": "Hong Kong",
    "slug": "hong-kong-hong-kong",
  },
]
`;
